yt_idefix/data_structures.py

```diff
@@ -427,15 +427,175 @@ def _get_time(self):
+                        1 / 3
+                    )
+
+        def _mass():
+            if "mass_unit" not in uo:
+                if "density_unit" in uo:
+                    uo["mass_unit"] = uo["density_unit"] * uo["length_unit"] ** 3
+                else:
+                    uo["mass_unit"] = (
+                        (uo["magnetic_unit"] / uo["velocity_unit"]) ** 2
+                        / 4.0
+                        / np.pi
+                        * uo["length_unit"] ** 3
+                    )
 
-        self.mass_unit = density_unit * self.length_unit ** 3
-        self.time_unit = self.length_unit / self.velocity_unit
-        self.magnetic_unit = np.sqrt(4.0 * np.pi * density_unit) * self.velocity_unit
+        def _velocity():
+            if "velocity_unit" not in uo:
+                if "length_unit" in uo and "time_unit" in uo:
+                    uo["velocity_unit"] = uo["length_unit"] / uo["time_unit"]
+                if "magnetic_unit" in uo:
+                    uo["velocity_unit"] = uo["magnetic_unit"] / np.sqrt(
+                        4.0 * np.pi * uo["density_unit"]
+                    )
+                else:
+                    uo["velocity_unit"] = (uo["mass_unit"] / uo["density_unit"]) ** (
+                        1 / 3
+                    ) / uo["time_unit"]
+
+        def _density():
+            if "density_unit" not in uo:
+                if "length_unit" in uo:
+                    uo["density_unit"] = uo["mass_unit"] / uo["length_unit"] ** 3
+                elif "velocity_unit" in uo:
+                    uo["density_unit"] = (
+                        (uo["magnetic_unit"] / uo["velocity_unit"]) ** 2 / 4.0 / np.pi
+                    )
+                else:
+                    uo["density_unit"] = (
+                        (uo["magnetic_unit"] * uo["time_unit"]) ** 3 / uo["mass_unit"]
+                    ) ** 2 / (4.0 * np.pi) ** 3
+
+        unit_deps = {
+            "length_unit": _length,
+            "mass_unit": _mass,
+            "time_unit": _time,
+            "velocity_unit": _velocity,
+            "density_unit": _density,
+        }
+
+        def cal_unit(unit):
+            """Derived units by recursion"""
+            try:
+                unit_deps[unit]()
+            except KeyError as err:
+                missed = err.args[0]
+                cal_unit(missed)
+                cal_unit(unit)
+
+        # Derive base units, the order doesn't matter
+        cal_unit("mass_unit")
+        cal_unit("length_unit")
+        cal_unit("time_unit")
+
+        self.mass_unit = uo["mass_unit"]
+        self.length_unit = uo["length_unit"]
+        self.time_unit = uo["time_unit"]
+
+        self.velocity_unit = self.length_unit / self.time_unit
+        self.density_unit = self.mass_unit / self.length_unit ** 3
+        self.magnetic_unit = (
+            np.sqrt(4.0 * np.pi * self.density_unit) * self.velocity_unit
+        )
         self.magnetic_unit.convert_to_units("gauss")
```

Comment:
@neutrinoceros Here's formula to derive magnetic, where density and velocity can be replaced with mass, time and length. There's no need of charge here because in cgs it's defined by mass, time and length

Reply:
Honestly speaking, I didn't derive the formula by myself but it's used in Pluto and also in other frontends such as athena and amrvac.

Reply:
In cgs, permittivity = permeability = 1 (dimensionless)

Reply:
Fair enough. I'll admit I've never sat through this derivation so I don't have a grasp of what "cgs" really means in the MHD simulation realm. I'll just trust you here as I did with other collaborators and frontends :)